Allocate IPv6 addresses after detecting IPv6 support

robmry opened the pull request:
**- What I did**

If IPv6 is disabled for a container, do not allocate an IPv6 address when it's attached to an IPv6 network.

- Fix https://github.com/moby/moby/issues/47055

**- How I did it**

There are a few commits - it's probably easiest to review them separately ...

_**Unconditionally update container.NetworkSettings**_

 In `daemon.allocateNetwork()`, flag `updateSettings` was set if `container.NetworkSettings.Networks` was empty. That flag was passed down through the call stack to avoid a call to `daemon.updateNetworkSettings()` in `daemon.updateNetworkConfig()`.

Other calls to `daemon.updateNetworkSettings()`, which happen when connecting an existing container to another network, unconditionally made the NetworkSettings update.

I don't think there's a circumstance where NetworkSettings is not empty during container creation. Even if there is, the update is harmless and cheap. Eliminating the `updateSettings` flag simplifies the code a little and, in the next commit where `allocateNetwork()` is split out of `initializeNetworking()`, it avoids the need to pass that flag between the two.

_**Separate Sandbox/Endpoint construction**_

Previously, the `libnetwork.Sandbox` was created in `daemon.allocateNetwork()` if there was no network - otherwise it happened while connecting an Endpoint in `daemon.connectToNetwork()`. If there was an endpoint in the default bridge, it had to be connected first, because extra configuration is needed in the Sandbox for legacy links and that could only happen during Sandbox construction.

Now, config for legacy links is added to the Sandbox when constructing the Endpoint that needs it - removing the constraint on ordering of Endpoint construction, and the dependency between Endpoint and Sandbox construction.

So, now a Sandbox can be constructed in one place, before the first Endpoint.

Also, replaces some legacy-link specific Sandbox option-setters with Sandbox methods for updating `/etc/hosts`, and updates a legacy-link parent's IPv6 address as well as its IPv4 address when a child container restarts.

_**Configure network endpoints after creating a container**_

This commit splits the bulk of `daemon.allocateNetwork()` out of `daemon.initializeNetworking()`.

`daemon.initializeNetworking()` hasn't moved, but it now only prepares configuration and doesn't do any Endpoint construction or configuration. Sandbox construction still happens here.

Endpoint construction is in `daemon.allocateNetwork()` which, on non-Windows, is called after the container task has been created (before it's started).

On Windows, Endpoint construction still happens before the container task is created. If it's created afterwards, some DNS lookups for the container don't seem to end up in our resolver - see the TODO comment in the code. (But, Windows can't benefit from the delayed assignment of IPv6 addresses anyway.)

Doing the Endpoint construction after the `osSbox` has been set up makes it possible to probe the container's network for IPv6 support first - enabling the next commit ...

_**Only assign IPv6 addresses if required**_

If a Sandbox is provided to `CreateEndpointForSandbox`, check whether it has IPv6 before allocating IPv6 addresses.

Because no IPv6 address is added to the DNS when the Sandbox doesn't support it, a lookup for a container on an IPv6 network can now have no IPv6 address. The resolver is updated to treat a missing IPv6 address on an IPv4 hit as `ipv6miss` even if the network has IPv6 enabled. (So, an empty DNS response is returned, avoiding the upstream DNS request and NXDOMAIN.)

The `macvlan` driver expected an endpoint on a network with IPv6 subnets to have an IPv6 address - it searched the subnets for the address assigned to the endpoint, to work out which gateway address to use (crashing if there was no address). Now, it only makes that search if there is an IPv6 address.

**- How to verify it**

Existing regression tests for the refactoring.

New integration test checks that a container with IPv6 disabled via `sysctl` on an IPv6 network has no IPv6 address.

**- Description for the changelog**

If IPv6 is disabled for a container, do not allocate an IPv6 address when it's attached to an IPv6 network.

robmry committed:
Configure network endpoints after creating a container

For Linux, delay construction and configuration of network endpoints
until the container has been created (but not started).

Signed-off-by: Rob Murray <rob.murray@docker.com>

corhere commented on daemon/network.go:
```diff
@@ -893,6 +893,13 @@ func buildCreateEndpointOptions(c *container.Container, n *libnetwork.Network, e
 
 	createOptions = append(createOptions, libnetwork.EndpointOptionGeneric(genericOptions))
 
+	// IPv6 may be enabled on the network, but disabled in the container.
+	if n.IPv6Enabled() {
+		if sbIPv6, ok := sb.IPv6Enabled(); ok && !sbIPv6 {
+			createOptions = append(createOptions, libnetwork.CreateOptionDisableIPv6())
+		}
+	}
```
The endpoint could determine for itself whether IPv6 is enabled in the sandbox when the Endpoint is Joined to the Sandbox. It'd make sense to have an endpoint option persisted in the store for _the user_ to disable IPv6 allocation for an attachment, but it feels weird to use an endpoint option to convey a fact about the sandbox it will be joined to. The endpoint option might get out of sync with ground truth.

How feasible would it be to defer IPAM allocation until the join operation?

robmry replied:
I think it's feasible - and I agree it'd be better, but it would be a behavioural change for libnetwork. At the moment, our daemon code expects to be able to stash address info in NetworkSettings after creating the endpoint, before the join. That can be changed but, do we need to worry about other users of libnetwork doing something similar?

Alternatively, we could pass the `libnetwork.Sandbox` to the `CreateEndpoint` as an `EndpointOption` - not to be stored in the `Endpoint`, just so it can see what it'll be joined to? (If it gets no sandbox option, or the sandbox doesn't have an osSbox that can be checked for the `::1`, the Endpoint would just follow its current rules for allocating IPv6 addresses - and libnetwork's behaviour wouldn't change.)

robmry replied:
> Alternatively, we could pass the libnetwork.Sandbox to the CreateEndpoint as an EndpointOption ...

Expanding on that idea, I added `CreateOptionJoinSandbox` ... nothing creates an Endpoint without adding it to a Sandbox. Optionally combining the two steps means the CreateEndpoint has access to the Sandbox, so it can inspect it.

robmry replied:
> > At the moment, our daemon code expects to be able to stash address info in NetworkSettings after creating the endpoint, before the join. That can be changed
> 
> Stash info for _which_ addresses? IPAM-allocated addresses are allocated without the daemon in the loop, IIUC. And I'm not sure why changing when IPAM addresses are allocated _inside libnetwork_ has any impact on daemon code.

The existing daemon code does ...
  - CreateEndpoint()
    - _libnetwork allocates addresses_
  - Copy the addresses libnetwork just allocated into NetworkSettings.
  - Join()

If libnetwork delayed addresses allocation until the Join(), NetworkSettings would have to be updated after the Join(). That's ok, but it is a change needed in the daemon because of an hidden change in the behaviour of libnetwork's API.

> > Optionally combining the two steps means the CreateEndpoint has access to the Sandbox, so it can inspect it.
> 
> CreateEndpoint would have access to _that instance_ of the Sandbox to inspect _the OS state at that point in time._ I'm worried about the restore story, where a container running when the daemon is stopped is restarted when the daemon is next started. Consider the case where the system is rebooted in between and `disable_ipv6=true` is added to or removed from the kernel command line. When the daemon is started, the existing Endpoint and Sandbox are rehydrated from libnetwork's datastore. If IPv6 support in the sandbox is only checked when the endpoint is created, the container would not get an IPv6 address if IPv6 support was enabled in the kernel between reboots, or an IPv6 address would be allocated that does not get used if IPv6 support was disabled. Actually, deferring IPAM allocation to join would not be sufficient to solve this restore case as sandbox restore follows a different code-path from endpoint join. IPAM allocation would also have to be redone on restore. (Redoing allocation on restore would also be required to e.g. properly implement a DHCP-client IPAM allocator.)

Yes, 'restore' restores a container with the config it had before. Re-creating a container and disconnect/connect are the ways to change its networking config.

Allocating addresses during Join(), and then not actually just Join() because that wouldn't help with the restore case anyway, would be a further complicated and delicate change - a lot of code paths are involved. So, while dealing with a requirement to change the host's networking stack without re-creating containers might be worthwhile - I think it implies significantly more risk and work, and doesn't need to be part of this change.

This PR cleans up construction of Sandbox and Endpoints by separating them cleanly, it doesn't move any responsibility for configuration of legacy links between the daemon and libnetwork, and it makes it possible to attach an IPv4 container to an IPv6 network without it getting IPv6 addresses.

robmry replied:
This also paves the way for not-allocating IPv4 addresses for an IPv6 only network, which is possibly more important.  There's quite a bit more to sort out for that, but this is a first step.

corhere replied:
I think it's safe to say there are no other users of libnetwork for container networking. Buildkit only uses the resolvconf package and any third-party users are doing so at their own risk, IMO.

robmry commented on libnetwork/sandbox.go:
```diff
@@ -273,6 +278,61 @@ func (sb *Sandbox) Refresh(options ...SandboxOption) error {
 
+// AddLegacyLinks configures legacy links, which are only supported by the default bridge network.
+func (sb *Sandbox) AddLegacyLinks(epAddrV4 string, children, parents map[string]*container.Container) error {
+	defaultNetName := runconfig.DefaultDaemonNetworkMode().NetworkName()
```
`AddLegacyLinks()` now just gets the `SandboxOption`s that would have been passed to the constructor... there's nothing to stop a caller passing in options that won't work properly post-construction - but I think the naming makes the intention clear enough.

robmry commented on libnetwork/sandbox.go:
```diff
@@ -273,6 +273,18 @@ func (sb *Sandbox) Refresh(options ...SandboxOption) error {
 	return nil
 }
 
+// AddLegacyLinks is used to configure legacy links, which are only supported by the default bridge network.
+func (sb *Sandbox) AddLegacyLinks(
+	extraHostsOptions []SandboxOption,
+	parentUpdateOptions []SandboxOption,
+	legacyLinksOption SandboxOption,
+) error {
+	sb.processOptions(extraHostsOptions...)
+	sb.processOptions(parentUpdateOptions...)
+	sb.processOptions(legacyLinksOption)
+	return sb.rebuildHostsFile()
+}
```
The option setters were only written to run before the constructor. In general, they won't work once the Sandbox has been created. The options needed to add a legacy link can be added later, but only before the Endpoint `Join` that needs them. Adding them after that just won't do anything.

In my proposed change ... libnetwork only sees the options it saw before, the daemon is still the only thing that has to calculate those options, and libnetwork is still the only thing that needs to know how to apply them.

The only thing both have to understand is when this method must be called. (And they only need to understand that until we delete legacy networks.)

A general-purpose-looking `ApplyOptions()` method would just be a new gotcha for anyone trying to work out how to set something in a Sandbox.

robmry replied:
Good point, that option should have been passed to the Join instead of Create.

corhere reviewed (commented):
If you are going to insist that IPAM must continue to happen when the endpoint is created, I would rather go back to the way you had it before where the daemon inspected the sandbox to decide whether or not to enable IPv6 on the endpoint. What you have currently is the same thing with many extra steps and lots of complexity without any upside.

Why is deferring IPAM allocation to Join/Restore infeasible and what do manually-assigned addresses have to do with it?

robmry committed:
Only allocate IPv6 addresses if IPv6 is enabled.

When a container doesn't support IPv6 and it's joined to an IPv6
network, don't allocate an IPv6 address for it.

Update the DNS resolver to understand that it can have an 'ipv6miss'
(meaning an IPv4 address exists, but no IPv6) when a network is
IPv6 enabled.

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry commented on libnetwork/endpoint.go:
```diff
@@ -96,6 +96,18 @@ type Endpoint struct {
+	optionDataAddLegacyLinks
+)
+
+// optionData is used by functional options to store data that does not belong
+// in the Endpoint itself.
+type optionData map[optionDataType]any
```
Yes, I started with a struct but because the same `EndpointOption` type is used for both create and join options, it meant having struct members that were only of use to one or the other. So, I tried this.

I don't like it either though, I'll go back to a struct.

robmry commented on libnetwork/endpoint.go:
```diff
@@ -64,7 +64,7 @@ func getNetworkType(nw *Network) int {
 // and Endpoint interfaces methods. The various setter functions of type EndpointOption are
 // provided by libnetwork, they look like <Create|Join|Leave>Option[...](...)
-type EndpointOption func(ep *Endpoint)
+type EndpointOption func(ep *Endpoint, od optionData)
```
It means the libnetwork user will need to recompile it, but that's what happens with Go (we're not shipping a compiled library), and there's no change in behaviour unless the new options are used.

The breaking change I was concerned about was a change in behaviour behind the scenes. Not detectable by a compiler, but suddenly no addresses are set up after CreateEndpoint(). That would surprise our daemon, containers would work but inspect wouldn't show addresses and they wouldn't be persisted. That's easily fixable, if you know the change has happened.

corhere replied:
```go
var foo libnetwork.EndpointOption = func(*Endpoint) {}
```
will no longer compile with this change. Therefore it is a breaking change. (I don't actually care; I'm just being pedantic.)

robmry committed:
Separate Sandbox/Endpoint construction

If config for legacy links needs to be added to a libnetwork.Sandbox,
add it when constructing the Endpoint that needs it - removing the
constraint on ordering of Endpoint construction, and the dependency
between Endpoint and Sandbox construction.

So, now a Sandbox can be constructed in one place, before the first
Endpoint.

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry commented:
I've un-merged CreateEndpoint/Endpoint.Join, so that the Sandbox can be configured for legacy links between the two.

CreateEndpoint still needs to know whether to set up IPv6 addresses, so I've added CreateEndpointForSandbox that takes the Sandbox as a param. Once we get rid of legacy links perhaps we'll make it do the Join as well, but there's no need to do that as part of this change.

Moving address assignment to Join would mean CreateEndpoint doesn't need to see the Sandbox, but it wouldn't help with legacy links ... Join would need a callback after allocating addresses, before doing the Join.

The final commit isn't intended to be merged in its current state, if we want to do it or something like it, the changes will need to be pushed down into the other commits - it's just an experiment/proposal for removing references to legacy-links from the libnetwork code, while making the Sandbox update for legacy links type safe.

> I do not want libnetwork to have any specific knowledge of containers or the daemon. The default bridge network is a daemon concept. Legacy links are a containers concept: containers have legacy links; sandboxes do not. The legacy links daemon feature happens to be implemented in terms of a combination of Sandbox options, but libnetwork does not need to know that.

That sounds good. But, the Sandbox options to implement legacy links are very legacy-linky, not generic things that we'll want to leave behind for other purposes when we drop support for legacy links. For example ...

```
// OptionParentUpdate function returns an option setter for parent container
// which needs to update the IP address for the linked container.
func OptionParentUpdate(cid string, name, ip string) SandboxOption {
```

Even `libnetwork.OptionGeneric` is only used for legacy links, but because it's disguised it's harder to reason about and the structure of the options is more complicated than it needs to be.

So, I don't think removing references to legacy links gives us anything but obfuscation. But, I know there are strong feelings on the subject - so, can do something like the change in the final commit? Or, is there a better approach?

corhere commented:
> CreateEndpoint still needs to know whether to set up IPv6 addresses, so I've added CreateEndpointForSandbox that takes the Sandbox as a param.

Please, please, **please** just make it the daemon's responsibility to pass `libnetwork.CreateOptionDisableIPv6()` to `CreateEndpoint`, same as the way you had it before I said anything. `func buildCreateEndpointOptions` already derives other endpoint-create options from the sandbox and everything. Keep in mind that I wanted the Endpoint's IPv6-ness to be determined _at Join time_ so that the Endpoint config could not get out of sync with the Sandbox. `CreateEndpointForSandbox` does not address that concern, _but superficially appears to do so_ because it takes a `*Sandbox` parameter.

--------

> But, the Sandbox options to implement legacy links are very legacy-linky, not generic things that we'll want to leave behind for other purposes when we drop support for legacy links.

Most of the Sandbox options to implement legacy links are very `/etc/hosts`-y. `OptionExtraHost` adds an entry to the sandbox's `/etc/hosts` file. It could just as easily be implemented as a `(*Sandbox)` method. `OptionParentUpdate` adds/updates an entry in another sandbox's `/etc/hosts` file. The meat of the implementation, `func (*Sandbox) updateParentHosts`, only uses its receiver parameter for two things: to get the list of updates to apply, and to get the controller handle so it can look up the other sandboxes to apply the updates to. If provided with a `(*Sandbox)` method to update a host entry, the daemon could just as easily implement that behaviour itself! It's already enumerating all the sandbox IDs which need updating and the entries to update anyway.

`OptionGeneric` is the only exception. The option data gets added to the data returned by `(*Sandbox).Labels()`, which in turn are passed into `(driverapi.Driver).Join` and `(driverapi.Driver).ProgramExternalConnectivity` because sandbox labels are endpoint driver options, obviously. (Sigh.) Oddly enough, the bridge driver respects the port binding and exposed ports options passed into `ProgramExternalConnectivity` but not `"ParentEndpoints"` or `"ChildEndpoints"`, despite the method reconfiguring parent/child link iptables rules.

`func (*Sandbox) AddLegacyLinks` therefore does three different things. It updates the /etc/hosts files of the receiver sandbox, it updates the /etc/hosts files of other sandboxes as specified by the `OptionParentUpdate` args, and it also updates the sandbox labels. The method call doesn't actually _do_ anything with the updated sandbox labels, however. Three distinct things should be three distinct functions.

```go
// strawman:

func (*Sandbox) AddHostsEntry(name, ip string) // instead of OptionExtraHost
func (*Sandbox) UpdateHostsEntry(name, ip string) // instead of OptionParentUpdate
func (*Sandbox) UpdateLabels(map[string]interface{}) // instead of OptionGeneric
```

There you go: type-safe and not obfuscated, just like you wanted. I'd even argue that it would _clarify_ legacy link behaviour since it makes explicit (on the daemon side) that a parent update is just an update to an entry in /etc/hosts of another sandbox. And it does not add any new references to legacy links to libnetwork, just like I wanted.

Note that with my strawman, the equivalent daemon code to
```go
libnetwork.OptionParentUpdate(parent.ID, alias, bridgeSettings.IPAddress)
```
would be (approximately)
```go
psb, _ := daemon.netController.GetSandbox(parent.ID)
psb.UpdateHostsEntry(alias, bridgeSettings.IPAddress)
```

`UpdateLabels` is still hard to reason about, but I don't see any better way that does not involve worse hacks or fundamentally rearchitecting core parts of libnetwork. There's no getting around the fact that the daemon has to do something else after updating the sandbox labels before the updated parent and child link lists take effect. Mentioning legacy links anywhere in the vicinity of such a method definition would imply behaviour it does not implement.

As for making `UpdateLabels` more type-safe, how about something like this?
```go
package bridge // import "github.com/docker/docker/libnetwork/drivers/bridge"

func LegacyContainerLinkOptions(parentEndpoints, childEndpoints []string) map[string]interface{}
```
```go
sb.UpdateLabels(bridge.LegacyContainerLinkOptions(parents, children))
```

I'm not picky about the name, but I feel pretty strongly about it living in the bridge driver package as the bridge driver is the only driver which consumes the parent/child endpoint lists.

robmry commented:
Ok - I think that's done.

corhere commented:
Thank you for humouring me! I skimmed the diff and it's looking much, much better. I am especially happy with how `func (*Daemon) addLegacyLinks` turned out! I will do a full review pass this week.

corhere commented on libnetwork/sandbox_unsupported.go:
```diff
@@ -30,3 +30,9 @@ func (sb *Sandbox) populateNetworkResources(*Endpoint) error {
+
+func (sb *Sandbox) IPv6Enabled() (enabled, ok bool) {
+	// Not implemented on Windows (Sandbox.osSbox is always nil).
+	// Assume the container supports IPv6, in case there's a Windows IPAM driver that does.
+	return true, true
```
Nope. https://github.com/moby/moby/blob/1f539a6e8581ae33f19d7e654404f8f0f74e36e1/libnetwork/drivers/windows/windows.go#L240-L242

```suggestion
	// None of the Windows container network drivers currently support IPv6.
	return false, true
```

robmry replied:
Ok.

corhere reviewed (changes requested):
Looking great, aside from a couple little details.

robmry commented on libnetwork/drivers/bridge/bridge_linux.go:
```diff
@@ -1078,27 +1078,29 @@ func (d *driver) CreateEndpoint(nid, eid string, ifInfo driverapi.InterfaceInfo,
 		if config.AddressIPv6 != nil {
 			network = config.AddressIPv6
 		}
+		// If there are fewer than 48 host-bits in the network, it's not possible to
+		// generate an IPv6 address - but the network size has already been checked if a
+		// MAC-based address is required.
```
That wouldn't work, because the driver isn't told to allocate an IPv6 address - it just does it if there isn't already an address and the network is IPv6-enabled. It's more legacy-networking / default-bridge fun ...

In `Network.createEndpoint()`, if the daemon has asked for `NetworkOptionDeferIPv6Alloc(true)`, `n.postIPv6` is set. Then, its first call to `assignAddress` [doesn't set up an IPv6 address](https://github.com/moby/moby/blob/330d777c53fbbf734178d6a35c9dc0a5070ba4ac/libnetwork/network.go#L1182) in the endpoint, even if one is required - giving this driver code a chance to invent one. After the driver's endpoint has been created, the driver-invented address is reserved by a second call to `assignAddress`.

The flag passed to `NetworkOptionDeferIPv6Alloc` is the result of a check that there are [enough bits in `fixed-cidr-v6`](https://github.com/moby/moby/blob/330d777c53fbbf734178d6a35c9dc0a5070ba4ac/daemon/daemon_unix.go#L1014-L1021) for the driver to generate the address.

So, before this change, in an IPv6 enabled network - if there wasn't already an IPv6 address it was because the daemon had arranged it that way.

Now, if there's no IPv6 address, it may be because the daemon is hoping the driver will invent an address for the default bridge - or it may be because the container doesn't support IPv6. In the latter case, there may be fewer than 48 host-bits, and that's not a problem.

robmry replied:
> If this is a special case for the default bridge network, should we fail loudly when those conditions are not met, e.g. ones > 80 && !n.postIPv6?

The network driver doesn't know about 'postIPv6'. Checking it in `libnetwork.Network` would mean libnetwork understanding when and why the option needs to be set - but that's the daemon's job.

I'll replace the comment with your equivalent text.

robmry committed:
Change meaning of return from DNSBackend.ResolveName

Interface DNSBackend.ResolveName, implemented by Network,
Sandbox (and noopDNSBackend) had a bool return value that
meant 'ipv6Miss'.

But, it was always set to true on a hit, and callers had
to deal with that.

So, changed the meaning of the return value to indicate
whether the name was found - which will also work for
'ipv4Miss' when we have IPv6-only containers/networks.

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry commented on libnetwork/network.go:
```diff
@@ -1986,6 +1990,7 @@ func (n *Network) ResolveName(ctx context.Context, req string, ipType int) ([]ne
 				ipLocal = append(ipLocal, net.ParseIP(ip.ip))
 			}
 		}
+		// TODO(robmry) - the bool return is 'ipv6Miss', it should be 'false' here.
```
`Sandbox.resolveName()` and `Sandbox.ResolveName()` also needed changes, it's a semantic change to `type DNSBackend interface`.

But, done (as a separate git commit).

robmry commented on daemon/container_operations_unix.go:
```diff
@@ -59,6 +62,87 @@ func (daemon *Daemon) setupLinkedContainers(container *container.Container) ([]s
+				}
+			}
+		}
+		if cEndpointID != "" {
+			parentEndpoints = append(parentEndpoints, cEndpointID)
+		}
```
Yes, that's silly ... but it's unrelated to this PR, which hasn't changed it at all.

robmry commented:
Moved back to draft for now ... the change will need to be rebased once the SetKey prestart hook is re-removed.

robmry committed:
Re-remove the SetKey OCI prestart hook

Second attempt to stop using the OCI prestart hook to call SetKey
to set up the OS Sandbox's key and perform network config in the
new network namespace.

The first attempt was reverted because it made it impossible to
use --sysctl to set per-interface sysctls on an interface that had
not yet been moved into the new network namespace.

Now, per-interface sysctls can be used to do that (with less
ambiguity because the setting is not tied to the interface using
an unpredictably assigned name).

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry committed:
Unconditionally update NetworkSettings

When connecting a container to a new network, its NetworkSettings were
unconditionally updated.

But, when creating a new container, they were only updated if there
were no NetworkSettings before a network was connected. But, that's
always the case - so, make the update unconditionally.

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry commented:
Rebased, and added a commit to re-remote the SetKey hook.

It'll need to be rebased again once https://github.com/moby/moby/pull/47621 (and that'll make `TestSetInterfaceSysctl` pass).

akerouanton commented on daemon/container_operations.go:
```diff
@@ -497,60 +438,34 @@ func (daemon *Daemon) allocateNetwork(ctx context.Context, cfg *config.Config, c
+		return err
 	}
 
-	// the intermediate map is necessary because "connectToNetwork" modifies "container.NetworkSettings.Networks"
```
I think it's worth keeping this comment.

akerouanton reviewed (commented):
Since we discussed on Slack about the attachment ordering issue in the 2nd commit, let me send my review as is. I'll take a look at other commits.

akerouanton commented on daemon/container_operations.go:
```diff
@@ -497,60 +438,34 @@ func (daemon *Daemon) allocateNetwork(ctx context.Context, cfg *config.Config, c
-	// on sandbox initialize for link, but the sandbox only be initialized
-	// on first network connecting.
-	defaultNetName := network.DefaultNetwork
-	if nConf, ok := ctr.NetworkSettings.Networks[defaultNetName]; ok {
```
After chatting more about this, I think you're right: the default gateway is changed whenever an endpoint that sorts first (in lexicographic order) is attached to the sandbox. We should be good to go with that change.

thaJeztah replied:
Marking as "unresolved" to keep these comments visible (making it easier to discover in future)

robmry replied:
Just to note then ... the endpoints are sorted by gateway-ness, then by name. (So, the "best" choices of gateway endpoints are preferred, but equally good endpoints are just ordered lexicographically.)

https://github.com/moby/moby/blob/59eba0ae1360171ba3b4fe36d0d35d9e8720ce90/libnetwork/sandbox.go#L677-L683

akerouanton commented on daemon/container_operations.go:
```diff
@@ -455,6 +482,28 @@ func (daemon *Daemon) allocateNetwork(ctx context.Context, cfg *config.Config, c
+func (daemon *Daemon) allocateNetwork(ctx context.Context, cfg *config.Config, ctr *container.Container) (retErr error) {
+	start := time.Now()
+
+	// An intermediate map is necessary because "connectToNetwork" modifies "container.NetworkSettings.Networks"
```
Oh, that comment was re-added in the 4th commit. You can forget about my previous comment.

robmry commented on libnetwork/sandbox_linux.go:
```diff
@@ -96,8 +96,12 @@ func (sb *Sandbox) updateGateway(ep *Endpoint) error {
-		return fmt.Errorf("failed to set IPv6 gateway while updating gateway: %v", err)
+	// If IPv6 has been disabled in the sandbox a gateway may still have been
+	// configured, don't attempt to apply it.
+	if ipv6, ok := sb.ipv6Enabled(); !ok || ipv6 {
```
Ok, yes - the function would have returned early if `osSbox == nil`, so the check is just unnecessary (until some change happens in the implementation `ipv6Enabled` that means it returns `!ok` for some other reason).

That code goes away again in a later commit, so I've changed it anyway.

robmry commented:
Rebased to resolve merge conflicts and run up-to-date tests.

robmry commented:
Rebased again (new conflict in `integration/networking/bridge_test_linux.go`).

thaJeztah commented on daemon/container_operations.go:
```diff
@@ -455,6 +482,28 @@ func (daemon *Daemon) allocateNetwork(ctx context.Context, cfg *config.Config, c
 		}
```
Not new code, and can't comment on the actual line above, but wondering if we should log if we fail to delete the sandbox.

robmry replied:
Ok, done.

thaJeztah commented on daemon/container_operations.go:
```diff
@@ -851,39 +900,6 @@ func (daemon *Daemon) normalizeNetMode(ctr *container.Container) error {
 	return nil
 }
 
-func (daemon *Daemon) initializeNetworking(ctx context.Context, cfg *config.Config, ctr *container.Container) error {
```
Looks like this patch swapped the order of `initializeNetworking` and  `allocateNetwork`, which makes the diff harder to read; would you mind changing the order back to reduce the diff?

robmry replied:
The functions weren't exactly swapped, code moved between them. I'd tried to minimise the diff in terms of lines-changed.

Maybe it's easier to understand a bigger diff, but without moving function declarations - just moving more lines between functions?

So I've made the change (can put it back if needed). The diff has gone from:
`1 file changed, 58 insertions(+), 42 deletions(-)` to
`1 file changed, 68 insertions(+), 52 deletions(-)`.

thaJeztah commented on daemon/start_linux.go:
```diff
@@ -5,30 +5,36 @@ import (
+		}
 	}
 	return nil
 }
```
Perhaps an early return here? That would also avoid having to do the `if err != nil` check for `allocateNetwork`;


```go
func (daemon *Daemon) initializeCreatedTask(ctx context.Context, cfg *config.Config, tsk types.Task, ctr *container.Container, spec *specs.Spec) error {
	if ctr.Config.NetworkDisabled {
		return nil
	}
	if nspath, ok := oci.NamespacePath(spec, specs.NetworkNamespace); ok && nspath == "" {
		// the runtime has been instructed to create a new network namespace for tsk.
		sb, err := daemon.netController.GetSandbox(ctr.ID)
		if err != nil {
			return errdefs.System(err)
		}
		if err := sb.SetKey(ctx, fmt.Sprintf("/proc/%d/ns/net", tsk.Pid())); err != nil {
			return errdefs.System(err)
		}
	}
	return daemon.allocateNetwork(ctx, cfg, ctr)
}
```

robmry replied:
Done.

thaJeztah commented on daemon/container_operations.go:
```diff
@@ -420,31 +421,57 @@ func (daemon *Daemon) updateContainerNetworkSettings(ctr *container.Container, e
+// initializeNetworking prepares network configuration for a new container.
+// If it creates a new libnetwork.Sandbox it's returned as newSandbox, for
+// the caller to Delete() if the container setup fails later in the process.
+func (daemon *Daemon) initializeNetworking(
+	ctx context.Context,
+	cfg *config.Config,
+	ctr *container.Container,
+) (newSandbox *libnetwork.Sandbox, retErr error) {
```
Slightly tempted to keep this unwrapped; yes, it's a long line, but most of it can be skimped over, knowing that it's just the func's signature.

robmry replied:
Done. Is the rule "no line-wrapping in function declarations"?

(To me, it's just slightly harder to read and see the return types. And diffs won't pick out future changes as clearly.)

akerouanton replied:
Most of the codebase is written that way, but I admit I'm not a big fan of that. To @robmry's point, it's harder to read IMHO.

thaJeztah replied:
Tempted to say; if the list of arguments for a function is becoming so long that it needs wrapping, that's a very good indicator that the signature may be in need of an overhaul 😄

robmry replied:
Two params and a return value plus context and error, seems ok? So we need shorter names?

robmry committed:
Make buildSandboxOption a function instead of a Daemon method

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry commented on daemon/container_operations.go:
```diff
@@ -497,60 +438,34 @@ func (daemon *Daemon) allocateNetwork(ctx context.Context, cfg *config.Config, c
-	// If the container is not to be connected to any network,
-	// create its network sandbox now if not present
-	if len(networks) == 0 {
-		if _, err := daemon.netController.GetSandbox(ctr.ID); err != nil {
```
Ok, done as a separate commit.

robmry commented on daemon/container_operations_unix.go:
```diff
@@ -59,6 +62,90 @@ func (daemon *Daemon) setupLinkedContainers(ctr *container.Container) ([]string,
+				return errors.Wrapf(err, "failed to add IPv6 address to /etc/hosts for link to %s", child.Name)
+			}
+		}
+		cEndpointID = defaultNW.EndpointID
```
That link doesn't work - but I think it's https://github.com/moby/moby/pull/47406#discussion_r1526578113

In any case, I think it's something for follow-up as it's not changed by this PR - which is already big enough!

thaJeztah commented on libnetwork/sandbox_windows.go:
```diff
@@ -37,3 +37,8 @@ func (sb *Sandbox) populateNetworkResources(context.Context, *Endpoint) error {
 	// not implemented on Windows (Sandbox.osSbox is always nil)
 	return nil
 }
+
+func (sb *Sandbox) IPv6Enabled() (enabled, ok bool) {
+	// None of the Windows container network drivers currently support IPv6.
```
This could be a GoDoc on the function, which may surface the information in IDEs etc. For example, my IDE (Goland) hides the function body by default in this case (hiding the comment 😬);

```go
// IPv6Enabled always returns false on Windows as None of the Windows container
// network drivers currently support IPv6.
func (sb *Sandbox) IPv6Enabled() (enabled, ok bool) {
	return false, true
}
```

robmry replied:
I'd never noticed GoLand hides comments like that!

Done - although now it's the only function in that file without a comment like that in its body.

thaJeztah commented on libnetwork/network.go:
```diff
@@ -1979,39 +1981,33 @@ func (n *Network) ResolveName(ctx context.Context, req string, ipType int) ([]ne
 	}
 
-	return nil, ipv6Miss
+	// this map is to avoid IP duplicates, this can happen during a transition period where 2 services are using the same IP
```
Mostly curious; is there any order in the results? In this case, should "last" or "first" take precedence? (currently the first one is used, and the last one is discarded; wondering if we should loop in reverse).

robmry replied:
This PR only changes whitespace here ... but no, I don't think the order can matter - the results come from a map in the first place, and DNS resolvers are free to rotate addresses in responses anyway to balance load.

thaJeztah commented on libnetwork/network.go:
```diff
@@ -1979,39 +1981,33 @@ func (n *Network) ResolveName(ctx context.Context, req string, ipType int) ([]ne
 
-	return nil, ipv6Miss
+	// this map is to avoid IP duplicates, this can happen during a transition period where 2 services are using the same IP
+	noDup := make(map[string]bool)
```
Not relevant for this PR, but this could uses a `map[string]struct{}` to avoid allocations.

thaJeztah reviewed (commented):
Did another pass; overall looking good. Perhaps ready to go, but more eyes won't hurt on this one!

robmry commented on daemon/container_operations.go:
```diff
@@ -667,10 +663,8 @@ func (daemon *Daemon) updateNetworkConfig(ctr *container.Container, n *libnetwor
+	if err := daemon.updateNetworkSettings(ctr, n, endpointConfig); err != nil {
+		return err
 	}
 	return nil
```
Ok.

robmry commented on daemon/container_operations_test.go:
```diff
@@ -20,6 +20,7 @@ func TestDNSNamesOrder(t *testing.T) {
 		Config: &containertypes.Config{
 			Hostname: "baz",
 		},
+		HostConfig: &containertypes.HostConfig{},
```
Yes ...

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x48 pc=0x17aaf44]

goroutine 27 [running]:
testing.tRunner.func1.2({0x1aa5ce0, 0x3171920})
	/usr/local/go/src/testing/testing.go:1631 +0x1c4
testing.tRunner.func1()
	/usr/local/go/src/testing/testing.go:1634 +0x33c
panic({0x1aa5ce0?, 0x3171920?})
	/usr/local/go/src/runtime/panic.go:770 +0x124
github.com/docker/docker/daemon.(*Daemon).updateNetworkSettings(0x400051dd00, 0x400051dab8, 0x40000504e0, 0x40001a5b20)
	/go/src/github.com/docker/docker/daemon/container_operations.go:170 +0x224
github.com/docker/docker/daemon.(*Daemon).updateNetworkConfig(0x400051dd00, 0x400051dab8, 0x40000504e0, 0x40001a5b20)
	/go/src/github.com/docker/docker/daemon/container_operations.go:631 +0x2ac
github.com/docker/docker/daemon.TestDNSNamesOrder(0x4000563d40)
	/go/src/github.com/docker/docker/daemon/container_operations_test.go:35 +0x264
testing.tRunner(0x4000563d40, 0x1ee8270)
	/usr/local/go/src/testing/testing.go:1689 +0xec
created by testing.(*T).Run in goroutine 1
	/usr/local/go/src/testing/testing.go:1742 +0x318
```

Which comes from:
https://github.com/moby/moby/blob/0cba41041441d1d44ed3915c64e14f3a8ba7d798/daemon/container_operations.go#L170

Looks like it's normally initialised in `daemon.newContainer` - and there's lots of code that assumes it exists:
https://github.com/moby/moby/blob/afdfe4ff86fa625ee38bbdadbfebe811efede3a9/daemon/container.go#L151

robmry commented on libnetwork/sandbox_dns_unix.go:
```diff
@@ -28,10 +28,24 @@ const (
-func (sb *Sandbox) finishInitDNS(ctx context.Context) error {
+// AddHostsEntry adds an entry to /etc/hosts.
+func (sb *Sandbox) AddHostsEntry(ctx context.Context, name, ip string) error {
+	sb.config.extraHosts = append(sb.config.extraHosts, extraHost{name: name, IP: ip})
+	return sb.rebuildHostsFile(ctx)
```
`Sandbox.AddHostsEntry` is only used for legacy `--link` options for containers on the default bridge network, before they're started - and those containers are only likely to be connected to the default bridge (only one Endpoint) ... so I don't think there's much to gain by optimising it.

`Sandbox.rebuildHostsFile` is called from `SetKey` for a newly created container once the its support for IPv6 is known. That container could be connected to multiple endpoints - so maybe there are some msecs to be gained by accumulating addresses from all of the endpoints. But, optimising the whole `etchosts` thing so that it doesn't keep writing/reading/merging the contents (before a container starts and the user's had a chance to modify it), is probably the way to go if that's an issue.

These `/etc/hosts` entries are pretty questionable anyway for user-defined networks that have an internal DNS resolver. If we figure out a [fix for buildkit](https://github.com/moby/moby/pull/48020), so that we can use the internal resolver for the default network too, even more so.

I don't really want to muck about with it in this PR, maybe as a follow-up if needed?

robmry replied:
This is now https://github.com/moby/moby/pull/48823

thaJeztah reviewed (approved):
LGTM

robmry commented on daemon/container_operations_unix.go:
```diff
@@ -59,6 +62,90 @@ func (daemon *Daemon) setupLinkedContainers(ctr *container.Container) ([]string,
+		return nil
+	}
+	for _, child := range children {
+		if !isLinkable(child) {
```
That's now https://github.com/moby/moby/pull/48821.
